Wikipedia/Code/WMFCVLInfo.m

```diff
@@ -0,0 +1,337 @@
+
+    
+    for (NSUInteger sectionIndex = 0; sectionIndex < numberOfSections; sectionIndex++) {
+        WMFCVLSection *section = nil;
+        if (sectionIndex >= [_sections count]) {
+            section = [WMFCVLSection sectionWithIndex:sectionIndex];
+            [_sections addObject:section];
+        } else {
+            section = _sections[sectionIndex];
+        }
+
+        currentColumn.width = [columnWeights[currentColumnIndex] doubleValue]*baselineColumnWidth;
+        CGFloat columnWidth = currentColumn.width;
+        
+        CGFloat x = contentInsets.left;
+        for (NSInteger i = 0; i < currentColumnIndex; i++) {
+            x += [columnWeights[i] doubleValue] * baselineColumnWidth + interColumnSpacing;
+        }
+        
+        if (sectionIndex == 0) {
+            currentColumn.height += contentInsets.top;
+        } else {
+            currentColumn.height += interSectionSpacing;
+        }
+        CGFloat y = currentColumn.height;
+        CGPoint sectionOrigin = CGPointMake(x, y);
+        
+        [currentColumn addSection:section];
+        
+        CGFloat sectionHeight = 0;
+        
+            NSIndexPath *supplementaryViewIndexPath = [NSIndexPath indexPathForRow:0 inSection:sectionIndex];
+        
+        CGFloat headerHeight = 0;
+        
+        if (section.headers.count == 0) {
+            headerHeight = [delegate collectionView:collectionView estimatedHeightForHeaderInSection:sectionIndex forColumnWidth:columnWidth];
+            WMFCVLAttributes *headerAttributes = (WMFCVLAttributes *)[WMFCVLAttributes layoutAttributesForSupplementaryViewOfKind:UICollectionElementKindSectionHeader withIndexPath:supplementaryViewIndexPath];
+            if (headerAttributes != nil) {
+                headerAttributes.frame = CGRectMake(x, y, columnWidth, headerHeight);
+                [section addHeader:headerAttributes];
+            }
+            [invalidatedHeaderIndexPaths addObject:supplementaryViewIndexPath];
+        } else {
+            WMFCVLAttributes *headerAttributes = section.headers[0];
+            headerHeight = headerAttributes.frame.size.height;
+            CGRect headerFrame =  CGRectMake(x, y, columnWidth, headerHeight);
+            if (!CGRectEqualToRect(headerFrame, headerAttributes.frame)) {
+                WMFCVLAttributes *newAttributes = [headerAttributes copy];
+                newAttributes.frame = headerFrame;
+                [section replaceHeaderAtIndex:0 withHeader:newAttributes];
+                [invalidatedHeaderIndexPaths addObject:supplementaryViewIndexPath];
+            }
+        }
+        
+        sectionHeight += headerHeight;
+        y += headerHeight;
+        
+        CGFloat itemX = x + sectionInsets.left;
+        CGFloat itemWidth = columnWidth - sectionInsets.left - sectionInsets.right;
+        for (NSInteger item = 0; item < [collectionView.dataSource collectionView:collectionView numberOfItemsInSection:sectionIndex]; item++) {
```

Comment:
Should we cache the results of `[collectionView.dataSource collectionView:collectionView numberOfItemsInSection:sectionIndex]` - is that called on each iteration of the loop?